File: server/src/main/java/com/kernith/easyinvoice/service/InvoiceItemService.java
package com.kernith.easyinvoice.service;

import com.kernith.easyinvoice.config.AuthPrincipal;
import com.kernith.easyinvoice.data.dto.invoiceitem.CreateInvoiceItemRequest;
import com.kernith.easyinvoice.data.dto.invoiceitem.UpdateInvoiceItemRequest;
import com.kernith.easyinvoice.data.model.Invoice;
import com.kernith.easyinvoice.data.model.InvoiceItem;
import com.kernith.easyinvoice.data.model.InvoiceStatus;
import com.kernith.easyinvoice.data.model.UserRole;
import com.kernith.easyinvoice.data.repository.InvoiceItemRepository;
import com.kernith.easyinvoice.data.repository.InvoiceRepository;
import com.kernith.easyinvoice.helper.Utils;
import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

/**
 * Manages invoice items and keeps invoice totals in sync.
 */
@Service
public class InvoiceItemService {

    private final InvoiceRepository invoiceRepository;
    private final InvoiceItemRepository invoiceItemRepository;

    /**
     * Creates the service with repositories.
     *
     * @param invoiceRepository invoice repository
     * @param invoiceItemRepository invoice item repository
     */
    public InvoiceItemService(InvoiceRepository invoiceRepository, InvoiceItemRepository invoiceItemRepository) {
        this.invoiceRepository = invoiceRepository;
        this.invoiceItemRepository = invoiceItemRepository;
    }

    /**
     * Adds a new item to an invoice and recalculates totals.
     *
     * <p>Lifecycle: validate role and invoice state, create item, save, then recalc totals.</p>
     *
     * @param invoiceId invoice identifier
     * @param request item creation payload
     * @param principal authenticated principal
     * @return saved invoice item
     * @throws ResponseStatusException if role or invoice state is invalid
     */
    public InvoiceItem addInvoiceItem(Long invoiceId, CreateInvoiceItemRequest request, AuthPrincipal principal) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER, UserRole.BACK_OFFICE));
        Invoice invoice = getEditableInvoice(invoiceId, principal);

        InvoiceItem item = new InvoiceItem(invoice);
        item.setPosition(request.position());
        item.setDescription(Utils.normalizeRequired(request.description()));
        item.setNotes(Utils.trimToNull(request.notes()));
        item.setQuantity(Utils.defaultBigDecimal(request.quantity(), BigDecimal.ONE));
        item.setUnit(Utils.trimToNull(request.unit()));
        item.setUnitPrice(Utils.defaultBigDecimal(request.unitPrice(), BigDecimal.ZERO));
        item.setTaxRate(Utils.defaultBigDecimal(request.taxRate(), BigDecimal.ZERO));
        item.setDiscountType(Utils.mapDiscountType(request.discountType()));
        item.setDiscountValue(Utils.defaultBigDecimal(request.discountValue(), BigDecimal.ZERO));

        InvoiceItem saved = invoiceItemRepository.save(item);
        recalcInvoiceTotals(invoice);
        return saved;
    }

    /**
     * Lists items for a given invoice.
     *
     * @param invoiceId invoice identifier
     * @param principal authenticated principal
     * @return list of invoice items
     * @throws ResponseStatusException if role or invoice state is invalid
     */
    public List<InvoiceItem> listInvoiceItems(Long invoiceId, AuthPrincipal principal) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER, UserRole.BACK_OFFICE));
        Invoice invoice = getInvoiceOrThrow(invoiceId, principal);
        return invoiceItemRepository.findByInvoiceIdOrderByPositionAsc(invoice.getId());
    }

    /**
     * Updates an existing invoice item and recalculates totals.
     *
     * <p>Lifecycle: validate role and invoice state, update fields, save, then recalc totals.</p>
     *
     * @param invoiceId invoice identifier
     * @param itemId item identifier
     * @param request update payload
     * @param principal authenticated principal
     * @return saved invoice item
     * @throws ResponseStatusException if role or invoice state is invalid
     */
    public InvoiceItem updateInvoiceItem(
            Long invoiceId,
            Long itemId,
            UpdateInvoiceItemRequest request,
            AuthPrincipal principal
    ) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER, UserRole.BACK_OFFICE));
        Invoice invoice = getEditableInvoice(invoiceId, principal);

        Optional<InvoiceItem> optionalItem = invoiceItemRepository.findByIdAndInvoiceId(itemId, invoice.getId());
        if (optionalItem.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Parameters not valid");
        }
        InvoiceItem item = optionalItem.get();

        if (request.position() != null) {
            item.setPosition(request.position());
        }
        if (request.description() != null) {
            item.setDescription(Utils.normalizeRequired(request.description()));
        }
        if (request.notes() != null) {
            item.setNotes(Utils.trimToNull(request.notes()));
        }
        if (request.quantity() != null) {
            item.setQuantity(request.quantity());
        }
        if (request.unit() != null) {
            item.setUnit(Utils.trimToNull(request.unit()));
        }
        if (request.unitPrice() != null) {
            item.setUnitPrice(request.unitPrice());
        }
        if (request.taxRate() != null) {
            item.setTaxRate(request.taxRate());
        }
        if (request.discountType() != null) {
            item.setDiscountType(Utils.mapDiscountType(request.discountType()));
        }
        if (request.discountValue() != null) {
            item.setDiscountValue(request.discountValue());
        }

        InvoiceItem saved = invoiceItemRepository.save(item);
        recalcInvoiceTotals(invoice);
        return saved;
    }

    /**
     * Deletes an invoice item and recalculates totals if found.
     *
     * @param invoiceId invoice identifier
     * @param itemId item identifier
     * @param principal authenticated principal
     * @return optional result indicating success
     * @throws ResponseStatusException if role or invoice state is invalid
     */
    public Optional<Boolean> deleteInvoiceItem(Long invoiceId, Long itemId, AuthPrincipal principal) {
        Utils.requireRoles(principal, List.of(UserRole.COMPANY_MANAGER, UserRole.BACK_OFFICE));
        Optional<Invoice> optionalInvoice = invoiceRepository.findByIdAndCompanyId(
                invoiceId,
                Utils.getRequiredCompanyId(principal)
        );
        if (optionalInvoice.isEmpty()) {
            return Optional.empty();
        }
        Invoice invoice = optionalInvoice.get();
        if (invoice.getStatus() != InvoiceStatus.DRAFT) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invoice is not editable");
        }

        Optional<InvoiceItem> optionalItem = invoiceItemRepository.findByIdAndInvoiceId(itemId, invoice.getId());
        if (optionalItem.isEmpty()) {
            return Optional.empty();
        }
        invoiceItemRepository.delete(optionalItem.get());
        recalcInvoiceTotals(invoice);
        return Optional.of(Boolean.TRUE);
    }

    private Invoice getInvoiceOrThrow(Long invoiceId, AuthPrincipal principal) {
        Long companyId = Utils.getRequiredCompanyId(principal);
        Optional<Invoice> optionalInvoice = invoiceRepository.findByIdAndCompanyId(invoiceId, companyId);
        if (optionalInvoice.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invoice is not editable");
        }
        return optionalInvoice.get();
    }

    private Invoice getEditableInvoice(Long invoiceId, AuthPrincipal principal) {
        Invoice invoice = getInvoiceOrThrow(invoiceId, principal);
        if (invoice.getStatus() != InvoiceStatus.DRAFT) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invoice is not editable");
        }
        return invoice;
    }

    private void recalcInvoiceTotals(Invoice invoice) {
        List<InvoiceItem> items = invoiceItemRepository.findByInvoiceIdOrderByPositionAsc(invoice.getId());
        invoice.recalculateTotalsFromItems(items);
        invoiceRepository.save(invoice);
    }

}

File: client/src/core/services/invoice-item-service.ts
import { BaseService } from "./base-service";

/** Payload for creating an invoice item. */
export interface CreateInvoiceItemRequest {
  /** Position index (1-based). */
  position: number;
  /** Item description. */
  description: string;
  /** Notes (optional). */
  notes?: string;
  /** Quantity. */
  quantity?: number;
  /** Unit (optional). */
  unit?: string;
  /** Unit price. */
  unitPrice?: number;
  /** Tax rate (optional). */
  taxRate?: number;
  /** Discount type (optional). */
  discountType?: "NONE" | "PERCENT" | "AMOUNT";
  /** Discount value (optional). */
  discountValue?: number;
}

/** Payload for updating an invoice item. */
export interface UpdateInvoiceItemRequest {
  /** Position index (1-based). */
  position?: number;
  /** Updated description. */
  description?: string;
  /** Updated notes. */
  notes?: string;
  /** Updated quantity. */
  quantity?: number;
  /** Updated unit. */
  unit?: string;
  /** Updated unit price. */
  unitPrice?: number;
  /** Updated tax rate. */
  taxRate?: number;
  /** Updated discount type. */
  discountType?: "NONE" | "PERCENT" | "AMOUNT";
  /** Updated discount value. */
  discountValue?: number;
}

/** Response payload for an invoice item. */
export interface InvoiceItemResponse {
  /** Item identifier. */
  id: number;
  /** Invoice identifier. */
  invoiceId: number;
  /** Position index (1-based). */
  position: number;
  /** Item description (optional). */
  description?: string;
  /** Notes (optional). */
  notes?: string;
  /** Quantity (optional). */
  quantity?: number;
  /** Unit (optional). */
  unit?: string;
  /** Unit price (optional). */
  unitPrice?: number;
  /** Tax rate (optional). */
  taxRate?: number;
  /** Discount type (optional). */
  discountType?: "NONE" | "PERCENT" | "AMOUNT";
  /** Discount value (optional). */
  discountValue?: number;
  /** Line subtotal amount. */
  lineSubtotalAmount?: number;
  /** Line tax amount. */
  lineTaxAmount?: number;
  /** Line total amount. */
  lineTotalAmount?: number;
}

/** Invoice item API service. */
export class InvoiceItemService extends BaseService {
  /** Base path for invoice item endpoints. */
  protected basePath(): string {
    return "/invoices";
  }

  /**
   * Adds an item to an invoice.
   * @param invoiceId invoice identifier
   * @param request item payload
   */
  addItem(invoiceId: number | string, request: CreateInvoiceItemRequest): Promise<InvoiceItemResponse> {
    return this.post<InvoiceItemResponse, CreateInvoiceItemRequest>(this.endpoint(`/${invoiceId}/items`), {
      body: request,
    });
  }

  /**
   * Lists items for an invoice.
   * @param invoiceId invoice identifier
   */
  listItems(invoiceId: number | string): Promise<InvoiceItemResponse[]> {
    return this.get<InvoiceItemResponse[]>(this.endpoint(`/${invoiceId}/items`));
  }

  /**
   * Updates an item by id.
   * @param invoiceId invoice identifier
   * @param itemId item identifier
   * @param request update payload
   */
  updateItem(
    invoiceId: number | string,
    itemId: number | string,
    request: UpdateInvoiceItemRequest
  ): Promise<InvoiceItemResponse> {
    return this.patch<InvoiceItemResponse, UpdateInvoiceItemRequest>(
      this.endpoint(`/${invoiceId}/items/${itemId}`),
      { body: request }
    );
  }

  /**
   * Deletes an item from an invoice.
   * @param invoiceId invoice identifier
   * @param itemId item identifier
   */
  deleteItem(invoiceId: number | string, itemId: number | string): Promise<void> {
    return this.delete<void>(this.endpoint(`/${invoiceId}/items/${itemId}`));
  }
}
